libs/pilight/protocols/433.92/rev_v5.c

```diff
@@ -0,0 +1,414 @@
+	10		STATE (MSB)
+	11		--- ONE ---
+	12		--- ONE ---
+LASTBIT
+
+STATE
+	0 => ALL in area ON
+	1 => UNIT        ON
+	2 => UNIT        OFF
+	3 => ALL in area OFF
+
+AREABIT
+	for Unit 0-3 the AREABIT is not set
+	for Unit 4-7 the AREABIT is set
+	Areabit is used to identify the units when the ALL state is used
+		units 0-3 (bit=0)
+		units 4-7 (bit=1)
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#include "../../core/pilight.h"
+#include "../../core/common.h"
+#include "../../core/dso.h"
+#include "../../core/log.h"
+#include "../protocol.h"
+#include "../../core/binary.h"
+#include "../../core/gc.h"
+#include "rev_v5.h"
+
+#define PULSE_MULTIPLIER	3
+#define NORMAL_REPEATS		10
+#define MIN_PULSE_LENGTH	410
+#define MAX_PULSE_LENGTH	490
+#define AVG_PULSE_LENGTH	400
+#define RAW_LENGTH		50
+
+#define ID_COUNT            	16
+
+static char idLetters[ID_COUNT] = {"ABIJCDKLGHOPEFMN"}; //A-P
+
+static int validate(void) {
+	if(rev5_switch->rawlen == RAW_LENGTH) {
+		if(rev5_switch->raw[rev5_switch->rawlen-1] >= (MIN_PULSE_LENGTH*PULSE_DIV) &&
+		   rev5_switch->raw[rev5_switch->rawlen-1] <= (MAX_PULSE_LENGTH*PULSE_DIV)) {
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+static void createMessage(int id, int unit, int state, int all) {
+	rev5_switch->message = json_mkobject();
+	json_append_member(rev5_switch->message, "id", json_mknumber(id, 0));
+	if(all >= 1) {
+		json_append_member(rev5_switch->message, "all", json_mknumber(all, 0));
+	} else {
+		json_append_member(rev5_switch->message, "unit", json_mknumber(unit, 0));
+	}
+
+	if(state == 0) {
+		json_append_member(rev5_switch->message, "state", json_mkstring("off"));
+	}
+	else if(state == 1) {
+		json_append_member(rev5_switch->message, "state", json_mkstring("on"));
+	}
+}
+
+static void parseCode(void) {
+	int x = 0, i = 0, binary[RAW_LENGTH/4];
+	int binaryUnit[2];
+	int id = 0; int unit = 0;
+	int areabit = 0;
+	int rawState = 0; int state = 0;
+	int all = 0;
+
+	if(rev5_switch->rawlen>RAW_LENGTH) {
+		logprintf(LOG_ERR, "rev5_switch: parsecode - invalid parameter passed %d", rev5_switch->rawlen);
+		return;
+	}
+
+	/* Convert the one's and zero's into binary */
+	for(x=0;x<rev5_switch->rawlen-2;x+=4) {
+		if(rev5_switch->raw[x+3] > (int)((double)AVG_PULSE_LENGTH*((double)PULSE_MULTIPLIER/2))) {
+			binary[i++] = 1;
+		} else {
+			binary[i++] = 0;
+		}
+	}
+
+	binaryUnit[0] = binary[4];
+	binaryUnit[1] = binary[7];
+	areabit = binary[6];
+
+	id = binToDec(binary, 0, 3);
+	unit = binToDec(binaryUnit, 0, 1);
+	if (areabit > 0) {
+		//areabit does group units in 0-3 / 4-7
+		unit += 4;
+	}
+
+	rawState = binToDec(binary, 8, 9);
+
+	//see createState()
+	if (rawState == 1) {
+		state = 1; //on
+		all = 0;
+	}
+	else if (rawState == 2) {
+		state = 0; //off
+		all = 0;
+	}
+	else if (rawState == 3) {
+		state = 0; //off
+		all = 1;
+	}
+	else if (rawState == 0) {
+		state = 1; //on
+		all = 1;
+	}
+
+	if (all > 0) {
+        //switch all units in group 1
+
```

Comment:
Bogus newline. And can you allow that comment?

I'd also rather see comments line this:
```
// switch all units in group 1
```
Or better
```
/*
 * Switch all units in group 1
 */
```

Reply:
done